✨ CAPI Operator: Add core provider conditions to API

alexander-demichev opened the pull request:
<!-- please add a icon to the title of this PR (see https://sigs.k8s.io/cluster-api/VERSIONING.md), and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patch and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:

This PR adds a preflight check for the core provider(described here https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20201020-capi-provider-operator.md#installing-a-provider). 

1. Added conditions to ProviderSpec.
2. Regenerated manifests.
3. Added a preflight check, it makes sure that no more than one CoreProvider CRs are created in any namespace.
4. Sets proper condition.
5. Added unit tests for all the above logic.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #
/area operator

alexander-demichev commented:
/retest

MarcelMue commented on exp/operator/api/v1alpha4/conditions_consts.go:
```diff
@@ -0,0 +1,31 @@
+	// exists in the cluster
+	MoreThanOneExistsReason = "MoreThanOneExists"
+
+	// MoreThanOneExistsReason a human readable message that explains PreflightCheckCondition
```
```suggestion
	// MoreThanOneExistsMessage a human readable message that explains PreflightCheckCondition
```

MarcelMue commented on exp/operator/controllers/coreprovider_controller_test.go:
```diff
@@ -0,0 +1,122 @@
+			},
+		},
+		{
+			name: "two core providers were created, preflight check failed",
```
Should have a third test case with two core providers in different namespaces.

MarcelMue commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,101 @@ package controllers
-func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
 	log := ctrl.LoggerFrom(ctx)
 	log.V(4).Info("Reconcile CoreProvider")
```
Can we remove this log line? Not really needed IMO.

MarcelMue commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,101 @@ package controllers
+		conditions.MarkFalse(coreProvider,
+			operatorv1.PreflightCheckCondition,
+			operatorv1.MoreThanOneExistsReason,
+			v1alpha4.ConditionSeverityInfo,
```
Why is the condition severity info? I would have expected error or warning.

alexander-demichev replied:
👍

alexander-demichev commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,101 @@ package controllers
+		return ctrl.Result{}, err
+	}
+
+	if len(coreProviderList.Items) > 1 {
```
I might be misunderstanding your suggestion but my understanding is that we are not allowing the deployment of more than one instance of CoreProvider, InfraProvider, etc. I'm expecting operator-managed deployment to behave this way: if the user creates a second instance of CR, the preflight check should fail.

> Last point. have you considered how we can avoid to re-implement the controller logic for each provider type, given that the reconcile logic is ~90% equal for all the provider types?

Yes, we discussed it last week with @MarcelMue. We'll probably create a library for shared functionality. For example logic for preflight checks can later be moved there.

fabriziopandini replied:
>  if the user creates a second instance of CR, the preflight check should fail.

I agree this is the expected behaviour, but I'm not sure `len>1 `will ensure that because 1 provider instance in the cluster  won't trigger preflight failure (len>1 is false).

> We'll probably create a library for shared functionality

I think we should get this effort started ASAP otherwise we are going to need review for the shared  functionalities multiple times (before and after move)

alexander-demichev replied:
You're right about reviewing the same things multiple times, I'll create a place for shared logic now.

alexander-demichev replied:
Regarding `len > 1`, I can rewrite the logic to explicitly look at name and kind.

alexander-demichev replied:
I gave it a bit of thinking, that's how I see the flow for preflight checks: 1. Get a list of all providers of a certain type. 2. If more than one exists, return a false condition. This should capture both kind and name check. 3. If only one provider is present move to the second preflight check(it is not implemented). 4. Return a true condition if all checks passed.

alexander-demichev commented:
@fabriziopandini @MarcelMue PTAL, a short summary of what the PR does:
1. Added conditions to ProviderSpec.
2. Regenerated manifests.
3. Added a preflight check, it makes sure that no more than one CoreProvider CRs are created in any namespace.
4. Sets proper condition.
5. Added unit tests for all the above logic.

JoelSpeed commented on exp/operator/api/v1alpha4/conditions_consts.go:
```diff
@@ -0,0 +1,31 @@
+	// exists in the cluster
+	MoreThanOneExistsReason = "MoreThanOneExists"
+
+	// MoreThanOneExistsMessage a human readable message that explains PreflightCheckCondition.
+	MoreThanOneExistsMessage = "More than one instance of core provider exists in the cluster"
```
This is currently specific to the core provider, do we need to include that in the name of the const so that we can have variations later for other types?

alexander-demichev replied:
This condition will be shared between providers, It's probably easier to remove "core" from the message.

JoelSpeed commented on exp/operator/api/v1alpha4/provider_types.go:
```diff
@@ -16,11 +16,17 @@ limitations under the License.
 // ProProviderStatus defines the observed state of the Provider.
-type ProviderStatus struct{}
+type ProviderStatus struct {
+	// Conditions define the current service state of the cluster.
```
Maybe update to provider?
```suggestion
	// Conditions define the current service state of the provider.
```

JoelSpeed commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,80 @@ package controllers
+		}
+	}()
+
+	// Ignore deleted CoreProvider, this can happen when foregroundDeletion
+	// is enabled
```
Maybe expand to explain why we don't need to clean up? Owner references will cause the GC to clean up all the resources created by the `CoreProvider` right?

alexander-demichev replied:
yeah, that's how I got it. https://kubernetes.io/docs/concepts/workloads/controllers/garbage-collection/#foreground-cascading-deletion

fabriziopandini replied:
There is an ongoing discussion about foreground deletion in https://github.com/kubernetes-sigs/cluster-api/issues/4085, and I think we should wait for this in order to address the problem consistently across the board.
Also, this relies on owner chain, and I'm not sure we are already at a stage where the chain of objects managed by each provider is yet fully defined.

JoelSpeed commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,60 @@
+func preflightChecks(ctx context.Context, c client.Client, gvk schema.GroupVersionKind) (*v1alpha4.Condition, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	log.V(4).Info("Performing flight checks.")
```
```suggestion
	log.V(4).Info("Performing preflight checks.")
```

JoelSpeed commented on exp/operator/controllers/shared_test.go:
```diff
@@ -0,0 +1,171 @@
+			gs.Expect(actualCondition.Status).To(Equal(tc.expectedCondition.Status))
+			gs.Expect(actualCondition.Severity).To(Equal(tc.expectedCondition.Severity))
+
+			for _, c := range tc.providers {
+				gs.Expect(testEnv.Delete(ctx, c)).To(Succeed())
+				// Wait for CoreProvider to be deleted.
+				g.Eventually(func() bool {
+					if err := testEnv.Get(ctx, client.ObjectKeyFromObject(c), c); err != nil {
+						if apierrors.IsNotFound(err) {
+							return true
+						}
+					}
+					return false
+				}, timeout).Should(BeTrue())
+			}
```
Can we use the `testEnv.Cleanup` to get rid of these? I'm concerned that having these after the expectations above will mean that they won't get cleaned up in case of failure.

Would be better to defer cleanup or deletion at the point each resource is created I think

alexander-demichev replied:
testEnv.Cleanup() doesn't seem to wait for object to be deleted

fabriziopandini commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,60 @@
+	// Checking if more than one instance of the same provider exists in the cluster.
+	if len(providerList.Items) > 1 {
+		log.V(4).Info("More than one core provider exists in the cluster. Only one is allowed.")
+		return &v1alpha4.Condition{
```
Please use `conditions.FalseCondition` to ensure consistency across the codebase

fabriziopandini commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,60 @@
+		}, nil
+	}
+
+	return &v1alpha4.Condition{
```
Please use conditions.TrueCondition (same as above)

fabriziopandini commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,60 @@
+	}
+
+	// Checking if more than one instance of the same provider exists in the cluster.
+	if len(providerList.Items) > 1 {
```
Sorry to go back to this point but I don't really grasp `Len()>1`

If there is already more than one provider of a given kind/name that's means that the cluster is already in a not supported state, so we are defeting the entire scope of preflight (prevents to get to a not supported state).

Also, we really need to check by kind/name, because it is perfectly acceptable to have two infrastructure providers (AWS and VSphere) or two bootstrap providers (kubeadm and Talos). What we should prevent instead is to have two instances of the same provider (e.g. two infrastructure AWS)

alexander-demichev replied:
Ok, I get it now. Thanks. We can get a list of all providers and check if there is no duplicate name. I think resource name is the only type indicator since it's also going to be used for fetching configuration.

JoelSpeed replied:
Do we not still want to have validation that there is only one `CoreProvider`, I get what is being said here for other kinds of providers but IIUC the `CoreProvider` is a singleton and cannot be duplicated? Maybe this is a case of the logic needing to be moved elsewhere?

alexander-demichev replied:
I updated the logic and added infrastructure providers to unit tests, PTAL

fabriziopandini commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,80 @@ package controllers
-		WithOptions(options).
-		Complete(r)
+func (r *CoreProviderReconciler) reconcile(ctx context.Context, coreProvider *operatorv1.CoreProvider) (ctrl.Result, error) {
+	// Run preflight checks to ensure that core provider can be installed properly
+	preflightCheckCondition, err := preflightChecks(ctx, r.Client, coreProvider.GroupVersionKind())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	conditions.Set(coreProvider, preflightCheckCondition)
+
+	// Requeue if preflight checks failed
+	if preflightCheckCondition.Status == v1.ConditionFalse {
+		return ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+	}
```
I think this could be simplified by having  preflightChecks to return `(ctrl.Result, error)`, which is also a pattern that we are using in other reconcilers to break down the reconcile logic in smaller parts

JoelSpeed commented on exp/operator/controllers/shared_test.go:
```diff
@@ -0,0 +1,253 @@
+			}
+
+			defer func() {
+				testEnv.Cleanup(ctx, tc.providers...)
```
What did we decide on cleanup? Does it work per test case or do we need to delete and wait?

alexander-demichev replied:
Each test creates its own resources, they should all be cleaned up in the end.

alexander-demichev commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,72 @@
+		}
+
+		// For any other provider we should check that instances with similar name exist in any namespace
+		if p.GetKind() != "CoreProvider" && p.GetName() == provider.GetName() {
+			log.V(4).Info("More than one provider exists in the cluster. Only one is allowed.")
+			return preflightFalseCondition, ctrl.Result{RequeueAfter: preflightFailedRequeueAfter}, nil
+		}
```
Right now the name is the only way of knowing provider type. We can check for the source of the providers once `FetchConfig` logic is implemented.

fabriziopandini replied:
TBH, given this is the first iteration, I don't think we should worries to much about people who intentionally want to trick the system.
Let's make sure this is documented in the KEP, eventually let's add a node in the book where clusterctl config is described, but once it is documented users should not configure two times the same provider with the same name, I think we are ok

fabriziopandini commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,72 @@
+
+	providerGVK := provider.GetObjectKind().GroupVersionKind()
+
+	providerList := &unstructured.UnstructuredList{}
```
Wondering if we can avoid to work with unstructured, because this makes the code difficult to read and to maintain.
All the components now shares the same spec and status; could we leverage on this to implement shared functions?

alexander-demichev replied:
I don't have any ideas right now. Maybe an internal type for serializing providers?

alexander-demichev replied:
You are right we can use shared spec/status for such functions, but this specific case looks at objects' Kind and Name.

fabriziopandini replied:
What about about creating a GenericProvider type in the controller package with TypeMeta, ObjectMeta, Spec and Status.
Each controller should convert to this type, trigger a common reconcileNormal/reconcileDelete funcs and then convert back to the original type before patching.
We can eventually also add helpers method on GenericProvider like isCoreProvider etc etc.

alexander-demichev replied:
👍 I've just added it, ptal

fabriziopandini commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,72 @@
+		}
+	}
+
+	return conditions.TrueCondition(operatorv1.PreflightCheckCondition), ctrl.Result{}, nil
```
I don't think we should return changes to apply to the object as a variable, because this pattern quickly became hard to maintain as soon as the number of changes grow.
Let update the object inside this function instead

alexander-demichev replied:
I tried to update the function, but it is not trivial with unstructured.

fabriziopandini commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,72 @@
+
+		// CoreProvider is a singleton resource, more than one instances should not exist
+		if p.GetKind() == "CoreProvider" {
+			log.V(4).Info("More than one core provider exists in the cluster. Only one is allowed.")
```
```suggestion
			log.V(4).Info("Core provider already exists in the cluster. Only one is allowed.")
```

fabriziopandini replied:
Also, given that we are able to distinguish a better log message for core providers, should this message surface to the user via conditions as well?

alexander-demichev replied:
👍

fabriziopandini commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,72 @@
+
+		// For any other provider we should check that instances with similar name exist in any namespace
+		if p.GetKind() != "CoreProvider" && p.GetName() == provider.GetName() {
+			log.V(4).Info("More than one provider exists in the cluster. Only one is allowed.")
```
```suggestion
			log.V(4).Info("There is already a %s with name %s in the cluster. Only one is allowed.", ...)
```

fabriziopandini commented on exp/operator/controllers/suite_test.go:
```diff
@@ -43,9 +49,20 @@ func TestAPIs(t *testing.T) {
 }
 
 func TestMain(m *testing.M) {
+	os.Setenv("TEST_ASSET_KUBECTL", "/tmp/kubebuilder/bin/kubectl")
+	os.Setenv("TEST_ASSET_KUBE_APISERVER", "/tmp/kubebuilder/bin/kube-apiserver")
+	os.Setenv("TEST_ASSET_ETCD", "/tmp/kubebuilder/bin/etcd")
```
I there a particular reason why we are enforcing this (we are not doing this in other tests...)

alexander-demichev replied:
no, some leftovers after my local debugging. Will remove it.

fabriziopandini commented on exp/operator/controllers/coreprovider_controller_test.go:
```diff
@@ -0,0 +1,80 @@
+
+var _ reconcile.Reconciler = &CoreProviderReconciler{}
+
+var _ = Describe("CoreProvider Reconciler", func() {
```
We agreed on using go test framework instead of ginkgo, see https://cluster-api.sigs.k8s.io/developer/testing.html for more context

fabriziopandini replied:
> I used the pattern that I saw here https://github.com/kubernetes-sigs/cluster-api/blob/master/controllers/machineset_controller_test.go#L42

I'm sorry you took as an example an test that is not yet aligned on the current recommendations.

> Are there plans to change other tests too?

Yep, /priority backlog /kind cleanup since the new recommendation are in place, but we see some progress, e.g. https://github.com/kubernetes-sigs/cluster-api/pull/4451

alexander-demichev commented on exp/operator/api/v1alpha4/conditions_consts.go:
```diff
@@ -0,0 +1,31 @@
+	MoreThanOneExistsReason = "MoreThanOneExists"
+
+	// MoreThanOneExistsMessage a human readable message that explains PreflightCheckCondition.
+	MoreThanOneExistsMessage = "More than one instance of the provider exists in the cluster"
```
Right, I can move it outside of consts file.

alexander-demichev commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,76 @@ package controllers
+	}
+
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
```
+1, can we do it in a follow-up? I can open an issue to track it.

alexander-demichev commented on exp/operator/api/v1alpha4/provider_types.go:
```diff
@@ -16,11 +16,17 @@ limitations under the License.
 
 // ProProviderStatus defines the observed state of the Provider.
-type ProviderStatus struct{}
+type ProviderStatus struct {
```
+1, a separate PR might be better, this one is getting too big.

alexander-demichev commented:
/retest

MarcelMue commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,87 @@ package controllers
-		For(&operatorv1alpha4.CoreProvider{}).
-		WithOptions(options).
-		Complete(r)
+func (r *CoreProviderReconciler) reconcile(ctx context.Context, coreProvider *genericProvider, unconvertedCoreProvider *operatorv1.CoreProvider) (_ ctrl.Result, reterr error) {
```
Why is unconvertedCoreProvider included here if it isn't needed currently?

JoelSpeed commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,87 @@ package controllers
+		return ctrl.Result{}, nil
+	}
+
+	// Convert provider to a generic type so it can use shared functions
+	convertedProvider, err := convertStructuredProviderToGeneric(coreProvider)
+	if err != nil {
+		return ctrl.Result{}, nil
+	}
```
Perhaps this could be done before the `patchHelper` defer so you can keep the deferred code into a single block? 

Also, bit of a nit, what about calling the var `genericProvider` rather than `convertedProvder`? I feel like it better represents what we are dealing with here

JoelSpeed commented on exp/operator/controllers/genericprovider.go:
```diff
@@ -0,0 +1,107 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// genericProvider a generic type for having ability to write shared function for providers
+type genericProvider struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   operatorv1.ProviderSpec   `json:"spec,omitempty"`
+	Status operatorv1.ProviderStatus `json:"status,omitempty"`
+}
```
Out of interest, did we consider the approach of using an interface rather than doing conversions? We could have used something like we do for the `Conditions` rather than having to perform type conversions

alexander-demichev replied:
Yes, I looked at this option, it will require implementing each function for each provider. If you have a strong opinion against conversions we can switch to interfaces.

JoelSpeed replied:
I don't have a strong opinion per se, but I do wonder if it is easier and more idiomatic to use interfaces (interfaces are how polymorphism works in Go right?). Also, maybe more in keeping with existing code here? @fabriziopandini Did you have any thoughts on conversion to internal type vs interface for this?

fabriziopandini replied:
IMO what we should avoid is to work with unstructured, which is complex and error prone; given that I'm open to any solution.
In this case, I'm not really sure we can use interfaces given that API types mostly expose fields vs exposing methods.

JoelSpeed replied:
My suggested interface would be 
```
type Provider interface {
  client.Object
  GetSpec() operatorv1.ProviderSpec
  SetSpec(in operatorv1.ProviderSpec)
  GetStatus() operatorv1.ProviderStatus
  SetStatus(in operatorv1.ProviderStatus)
}
```

Implementing these 4 simple functions on each of the 4 types of providers would mean that they can all be passed into the functions that need to be generic and instead of reading the spec from the object we would `provider.GetSpec().field`, so it's a simple change without having to do any conversion

This is very much inspired by the Conditions interface that we have across CAPI today

JoelSpeed commented on exp/operator/controllers/genericprovider.go:
```diff
@@ -0,0 +1,107 @@
+	g.Status.Conditions = conditions
+}
+
+// DeepCopyObject this is needed only for implementing an interface in order to be able
+// to use conditions library
+func (g *genericProvider) DeepCopyObject() runtime.Object {
+	return nil
+}
```
Does the condition setting actually work when you don't have this implemented? 🤔  I wonder if there are unknown consequences of this just returning nil. Is there a way we can actually implement this?

alexander-demichev replied:
I didn't find any issues it can cause and yes condition setting works.

fabriziopandini replied:
I'm +1 to implement this, so we avoid any possible side effect.

JoelSpeed commented on exp/operator/controllers/genericprovider_test.go:
```diff
@@ -0,0 +1,146 @@
+	g.Expect(err).ToNot(HaveOccurred())
+	g.Expect(reflect.DeepEqual(genericProvider.ObjectMeta, coreProvider.ObjectMeta)).To(BeTrue())
+	g.Expect(reflect.DeepEqual(genericProvider.TypeMeta, coreProvider.TypeMeta)).To(BeTrue())
+	g.Expect(reflect.DeepEqual(genericProvider.Status.Conditions, coreProvider.Status.Conditions)).To(BeTrue())
```
Do we not need to check the Spec and entire Status? (This applies to other tests too)

JoelSpeed commented on exp/operator/controllers/shared.go:
```diff
@@ -0,0 +1,79 @@
+		)
+
+		// CoreProvider is a singleton resource, more than one instances should not exist
+		if p.Kind == "CoreProvider" {
```
Can we make a const for `CoreProvider` since this is used a few times

alexander-demichev commented on exp/operator/api/v1alpha4/conditions_consts.go:
```diff
@@ -0,0 +1,28 @@
+limitations under the License.
+*/
+
+package v1alpha4
```
I'll create an issue to track and move the version to v1aplha1.

fabriziopandini commented on exp/operator/controllers/coreprovider_controller.go:
```diff
@@ -19,31 +19,85 @@ package controllers
-func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
-	log := ctrl.LoggerFrom(ctx)
-	log.V(4).Info("Reconcile CoreProvider")
+func (r *CoreProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (_ ctrl.Result, reterr error) {
```
Wondering if we can avoid 4 reconcilers. What if we have a 

```go
type GenericProviderReconciler struct {
	Object client.Object
	Client client.Client
}

func (r *GenericProviderReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager, options controller.Options) error {
	return ctrl.NewControllerManagedBy(mgr).
		For(r.Object).
		WithOptions(options).
		Complete(r)
}

func (r *GenericProviderReconciler) Reconcile(ctx context.Context, req reconcile.Request) (_ reconcile.Result, reterr error) {
	typedProvider := r.Object.DeepCopyObject()
	if err := r.Client.Get(ctx, req.NamespacedName, typedProvider); err != nil {
		if apierrors.IsNotFound(err) {
			// Object not found, return. Created objects are automatically garbage collected.
			// For additional cleanup logic use finalizers.
			return ctrl.Result{}, nil
		}
		// Error reading the object - requeue the request.
		return ctrl.Result{}, err
	}

	// Convert provider to a generic type so it can use shared functions
	genericProvider, err := r.toGenericProvider(typedProvider)

	// Initialize the patch helper
	patchHelper, err := patch.NewHelper(typedProvider, r.Client)
	if err != nil {
		return ctrl.Result{}, err
	}

	defer func() {
		if err := patchHelper.Patch(ctx, r.toTypedProvider(typedProvider)); err != nil {
			reterr = kerrors.NewAggregate([]error{reterr, err})
		}
	}()
	
	...
}
```

r.Object will gives us all the info to do conversion when necessary...

alexander-demichev replied:
thanks for idea!

fabriziopandini commented on exp/operator/controllers/genericprovider.go:
```diff
@@ -0,0 +1,107 @@
+	return nil
+}
+
+func listGenericProviders(ctx context.Context, c client.Client, gvk schema.GroupVersionKind) ([]genericProvider, error) {
```
This is falling back on unstructured.
Wondering if we can this by relying on r.Object and then using t.toGenericProvider

fabriziopandini reviewed (commented):
@alexander-demichev I like the direction it is taking and if possible I would like to push the context a little bit further by having a genricReconciler that is aware of the object is reconciling and adapt to this by avoiding fully avoiding unstructured and serialization/deserialization for conversions.

alexander-demichev commented on exp/operator/controllers/genericprovider.go:
```diff
@@ -0,0 +1,107 @@
+}
+
+func convertStructuredProviderToGeneric(provider client.Object) (*genericProvider, error) {
+	unstructuredProvider, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&provider)
```
I don't think that spec and status are accessible directly here.

JoelSpeed replied:
Yeah they won't be as we don't have a concrete type, we have an interface. I think our options here are either to use the converter as we have already or to create our own interface for providers as suggested https://github.com/kubernetes-sigs/cluster-api/pull/4520#discussion_r639802937

JoelSpeed commented on exp/operator/api/v1alpha4/bootstrapprovider_types.go:
```diff
@@ -50,6 +76,15 @@ type BootstrapProviderList struct {
 
+func (b *BootstrapProviderList) GetItems() []Provider {
+	providers := []Provider{}
+	for _, provider := range b.Items {
+		providers = append(providers, &provider)
+	}
```
You are taking the location of a loop variable here, this will cause issues, I think you need to copy the object before you put it into the list else the list will be full of the same value over and over.

You can see what I mean with this playground: https://play.golang.org/p/JAy1S1N1J5X

Have you got a test at all that will prove that this doesn't copy the same item from the list over and over?
```suggestion
	for _, provider := range b.Items {
	        p := provider
		providers = append(providers, &p)
	}
```
Will need to update this for all the different types

JoelSpeed reviewed (commented):
I think this looks good and makes the code a lot simpler, there's a few gotchas I think we need to account for though, please see my comments inline

JoelSpeed commented on exp/operator/controllers/genericprovider_controller.go:
```diff
@@ -0,0 +1,89 @@
+)
+
+type GenericProviderReconciler struct {
+	Provider     operatorv1.Provider
```
@fabriziopandini I'd like your opinion on this as I think you had thought about this already when you suggested the generic reconcile

fabriziopandini replied:
Definitely we should not reuse the same object.

However I think we should get something like
```
type GenericProviderReconciler struct { 
    Provider client.Object
}
```
And then when we can have:
```
func (r *GenericProviderReconciler) NewGenericPriovider() GenericProvider { 
    // check r. Provider and return a Wrapper for a new instance of it, with the wrapper implementing the GenericProvider interface
}
```

Similarly we can have a `ListGenericPrioviders(Provider client.Object)`, that gets the right typed list and then converts all the items into the corresponding GenericProvider Wrapper

Note 1. having 4 provider wrappers is a little bit more verbose than implementing those methods directly in the API, but given that the Generic thing is an implementation detail I don't want to surface it in the API.
Note 2. having type switches is not ideal, but given that we have only 4 provider types I consider this an acceptable trade off vs exposing implementation details in the API

fabriziopandini commented on exp/operator/api/v1alpha4/interfaces.go:
```diff
@@ -0,0 +1,39 @@
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// +k8s:deepcopy-gen=false
+type Provider interface {
+	client.Object
+	conditions.Setter
+	GetSpec() ProviderSpec
+	SetSpec(in ProviderSpec)
+	GetStatus() ProviderStatus
+	SetStatus(in ProviderStatus)
+}
+
+// +k8s:deepcopy-gen=false
+type ProviderList interface {
+	client.ObjectList
+
+	GetItems() []Provider
+}
```
I agree we should get those type, however I think we should keep the API surface as clean as possible and avoid to surface controller implementation details at this level.

What about moving those types under controllers / another package?
This will require also to have wrapper types / factory methods instead of implementing the interface methods in the API (more down in the review about this idea)

JoelSpeed commented on exp/operator/controllers/preflight_checks.go:
```diff
@@ -0,0 +1,82 @@
+		)
+
+		// CoreProvider is a singleton resource, more than one instances should not exist
+		if p.GetObjectKind().GroupVersionKind().Kind == coreProvider {
```
Do we need to use the GVK info here or can we use the type info, eg `p.GetObject()` and attempt to type assert it, I'm thinking you could do a little `isCoreProvider` util

```
func isCoreProvider(p genericProvider.GenericProvider) bool {
    _, ok := p.GetObject().(*providerv1.CoreProvider)
    return ok
}
```

JoelSpeed reviewed (commented):
Looks good for the most part, added some suggestions that might make the code a bit cleaner, LMK what you think

alexander-demichev commented on exp/operator/controllers/genericprovider_controller_test.go:
```diff
@@ -0,0 +1,210 @@
+			} else {
+				g.Expect(reflect.TypeOf(genericProvider)).To(Equal(reflect.TypeOf(tc.expectedType)))
+				g.Expect(err).ToNot(HaveOccurred())
+			}
```
should probably use `BeIdenticalTo` instead of `Equal` here

JoelSpeed commented on exp/operator/controllers/preflight_checks.go:
```diff
@@ -0,0 +1,83 @@
+		}
+
+		// For any other provider we should check that instances with similar name exist in any namespace
+		if p.GetObjectKind().GroupVersionKind().Kind != coreProvider && p.GetName() == provider.GetName() {
```
Can use the util `IsCoreProvider` here too, just negate it

alexander-demichev replied:
👍

JoelSpeed commented on exp/operator/controllers/genericprovider_controller.go:
```diff
@@ -0,0 +1,137 @@
+func (r *GenericProviderReconciler) NewGenericProvider() (genericprovider.GenericProvider, error) {
+	switch provider := r.Provider.(type) {
+	case *operatorv1.CoreProvider:
+		return &genericprovider.CoreProviderWrapper{CoreProvider: provider.DeepCopy()}, nil
```
Does a `DeepCopy` produce the right type here? Do we need to care? Any reason not to construct a fresh one, `&operatorv1.CoreProvider{}`, it would be more efficient

alexander-demichev replied:
👍

JoelSpeed reviewed (commented):
/lgtm

Thanks Alex! 

/assign @fabriziopandini for final review

fabriziopandini commented:
/lgtm
I will give few days for the other reviewers to take a look, then I will approve 
/cc @vincepri or @sbueringer @CecileRobertMichon @enxebre PTAL

fabriziopandini commented:
I don't see other people chiming in, soo
/approve

fabriziopandini commented:
/retest

sbueringer commented:
> I don't see other people chiming in now, soo let's keep this effort moving (we will eventually iterate later on)
> /approve

I'll try to review the upcoming operator PRs ;)
